api/spec/src/entitlements/entitlements.tsp

```diff
@@ -0,0 +1,319 @@
+}
+
+/**
+ * Type of the entitlement.
+ */
+@friendlyName("EntitlementType")
+enum EntitlementType {
+  metered: "metered",
+  boolean: "boolean",
+  static: "static",
+}
+
+// NOTE: use OmitProperties<T> for templates once entitlement models are defined
+
+/**
+ * Entitlement templates are used to define the entitlements of a plan.
+ * Features are omitted from the entitlement template, as they are defined in the rate card.
+ */
+@discriminator("type")
+@friendlyName("Entitlement")
+union Entitlement {
+  metered: EntitlementMetered,
+  static: EntitlementStatic,
+  boolean: EntitlementBoolean,
+}
+
+/**
+ * Shared fields for entitlement creation
+ */
+@friendlyName("EntitlementCreateSharedFields")
+model EntitlementCreateSharedFields {
+  /**
+   * The feature the subject is entitled to use.
+   * Either featureKey or featureId is required.
+   */
+  @example("example-feature-key")
+  featureKey?: Key;
+
+  /**
+   * The feature the subject is entitled to use.
+   * Either featureKey or featureId is required.
+   */
+  @example("01ARZ3NDEKTSV4RRFFQ69G5FAV")
+  featureId?: ULID;
+
+  /**
+   * Additional metadata for the feature.
+   */
+  metadata?: Metadata;
+
+  usagePeriod?: RecurringPeriod;
+}
+
+/**
+ * Create inpurs for metered entitlement
+ */
+@friendlyName("EntitlementMeteredCreateInputs")
+model EntitlementMeteredCreateInputs {
+  ...OmitProperties<EntitlementCreateSharedFields, "usagePeriod">;
+  type: EntitlementType.metered;
+
+  /**
+   * If softLimit=true the subject can use the feature even if the entitlement is exhausted, hasAccess will always be true.
+   */
+  isSoftLimit?: boolean = false;
+
+  /**
+   * Deprecated, ignored by the backend. Please use isSoftLimit instead; this field will be removed in the future.
+   */
+  @deprecated("Use isSoftLimit instead")
+  isUnlimited?: boolean = false;
+
+  usagePeriod: RecurringPeriod;
+  measureUsageFrom?: MeasureUsageFrom;
+
+  /**
+   * You can grant usage automatically alongside the entitlement, the example scenario would be creating a starting balance.
+   * If an amount is specified here, a grant will be created alongside the entitlement with the specified amount.
+   * That grant will have it's rollover settings configured in a way that after each reset operation, the balance will return the original amount specified here.
+   * Manually creating such a grant would mean having the "amount", "minRolloverAmount", and "maxRolloverAmount" fields all be the same.
+   */
+  @minValue(0)
+  issueAfterReset?: float64;
+
+  /**
+   * Defines the grant priority for the default grant.
+   */
+  @minValue(1)
+  @maxValue(255)
+  issueAfterResetPriority?: uint8 = 1;
+
+  /**
+   * If true, the overage is preserved at reset. If false, the usage is reset to 0.
+   */
+  preserveOverageAtReset?: boolean = false;
+}
+
+/**
+ * Create inputs for static entitlement
+ */
+@friendlyName("EntitlementStaticCreateInputs")
+model EntitlementStaticCreateInputs {
```

Comment:
```suggestion
model EntitlementStaticCreateInputs extend EntitlementCreateSharedFields {
```

Reply:
same question as above

Reply:
`allOf` and types can be reused in the underlying server, client code.